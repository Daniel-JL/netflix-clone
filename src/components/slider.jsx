import React, { useState, useRef, useEffect } from 'react';
import {
  NavLink,
} from 'react-router-dom';
import styled from 'styled-components';
import { SliderItem } from './slider-item';
import { ArrowButton } from './buttons';
import { useFetch } from '../hooks/useFetch';

const SliderContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`;

const ArrowButtonContainer = styled.div`
  height: 140px;
  width: 50px;
`;

export function Slider(props) {
  const [paginationPage, setPaginationPage] = useState(0);
  const [buttonHasBeenPressed, setButtonHasBeenPressed] = useState(false);
  const [numOfVisibleSliderItems, setNumOfVisibleSliderItems] = useState(6);
  const [dataLoaded, setDataLoaded] = useState(false);
  let movieTvShowIds = [];

  const mediaName = [];
  const imageId = [];
  const maturityRating = [];
  const numOfSeasons = [];

  const processSliderItemData = () => {

  };

  useEffect(() => {
    //  TODO
    //  1. Have a way of checking if we need to load data on state change
    //  2. If we need to get data then fetch it and process it
    if (!dataLoaded) {
      movieTvShowIds = fetchSliderItemIds();
      console.log(movieTvShowIds);
    }
  }, []);

  return (
    //  TODO
    //  Get a bunch of movie deets and then pass props to
    //  SliderItems so they can grab media they need
    <SliderContainer>
      <ArrowButtonContainer />
      <SliderItem />
      <SliderItem />
      <SliderItem />
      <SliderItem />
      <SliderItem />
      <SliderItem />
      <ArrowButtonContainer>
        <ArrowButton />
      </ArrowButtonContainer>
    </SliderContainer>
  );
}

export const fetchSliderItemIds = () => {
  const maxIdsNeeded = 42;
  const itemsPerPage = 20;
  let page = 1;
  const idArr = [];

  const fetchNow = () => {
    fetch(`https://api.themoviedb.org/3/trending/all/week?page=${page}&api_key=${process.env.REACT_APP_MOVIE_DB_API_KEY}`)
      .then((response) => response.json())
      .then((data) => {
        for (let i = 0; i < itemsPerPage; i++) {
          if (mediaIsMovieOrTv(data.results[i].media_type)) {
            idArr.push(data.results[i].id);
          }

          if (idArr.length >= maxIdsNeeded) {
            break;
          }
        }

        if (idArr.length < maxIdsNeeded) {
          page++;
          fetchNow();
        }
      });
  };
  fetchNow();
  return idArr;
};

const mediaIsMovieOrTv = (item) => (item === 'movie' || item === 'tv');
