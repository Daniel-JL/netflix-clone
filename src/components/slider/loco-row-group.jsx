import React, { useState, useRef, useEffect } from 'react';
import styled from 'styled-components';
import LoadingSkeleton from '../loading-skeleton';
import { MotionBackground } from '../motion-background/motion-background';
import LocoRow from './loco-row';

const GroupContainer = styled.div`
  position: relative;
  display: none;

  ${({ itemsLoaded }) => itemsLoaded && `
      display: block;
  `}
`;

const ContentContainer = styled.div`

`;

const ScrollLoadPoint = styled.div`
  position: absolute;
  width: 100%;
  height: 50%;
  top: 50%;
`;

const LocoRowGroup = ({
  genreTypeArr, 
  trendingItemData,
  numItems,
  setModalProps,
  numSlidersLoaded,
  setInfiniteScrollLoadPoint,
  portalRef,
}) => {
  const [itemsLoaded, setItemsLoaded] = useState(false);
  const numItemsLoaded = useRef(0);
  let scrollRef = useRef(0);
  let itemsToLoad;


  if (numSlidersLoaded === 1) {
    itemsToLoad = 5;
  } else {
    itemsToLoad = 4;
  }

  const handleItemLoaded = () => {
    console.log('handleItemLoaded');
    numItemsLoaded.current += 1;
    
    if (numItemsLoaded.current > itemsToLoad) {
      setItemsLoaded((itemsLoaded) => itemsLoaded + 1);
    }
  };

  return (
    <div>
      {!itemsLoaded && <LoadingSkeleton />}
      <GroupContainer itemsLoaded={itemsLoaded}>
        {itemsLoaded && 
          <ScrollLoadPoint 
            id="scrollLoadPoint" 
            ref={setInfiniteScrollLoadPoint}
          />
        }
        <ContentContainer>
          {numSlidersLoaded === 1 &&
            <MotionBackground
              mediaType={trendingItemData.mediaType}
              mediaId={trendingItemData.id}
              ageRating={trendingItemData.ageRating}
              isEpsInfoBox={false}
              handleItemLoaded={handleItemLoaded}
              itemsLoaded={itemsLoaded}
              portalRef={portalRef}
            />
          }
          {
            [
              ...Array(numItems),
            ].map((value: undefined, index: number) => (
              <LocoRow 
                key={index + numSlidersLoaded}
                genreName={genreTypeArr[index].genre.name}
                genreId={genreTypeArr[index].genre.id}
                mediaType={genreTypeArr[index].mediaType}
                numSlidersLoaded={index + 1 + numSlidersLoaded}
                setModalProps={setModalProps}
                handleItemLoaded={handleItemLoaded}
                itemsLoaded={itemsLoaded}
              />
            ))
          }
        </ContentContainer>
        
        
      </GroupContainer>
    </div>
  )
};

export default LocoRowGroup;
