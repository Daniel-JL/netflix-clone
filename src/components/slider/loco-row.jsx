import React, { useState } from 'react';
import styled from 'styled-components';
import LoadingSkeleton from '../loading-skeleton';
import MediaSliderContainer from './media-slider-container';

const RowContainer = styled.div`

`;

const RowPadding = styled.div`
  width: 100%;
  height: 3vw;
`;

const RowHeader = styled.header`
  width: 100%;
  font-size: 2vw;
  font-weight: bold;
  height: 2vw;
  color: white;
  z-index: 4;

`;

const HeaderText = styled.div`
  padding-left: 4.5vw;
`;

const Row = styled.div`
  position: relative;
  z-index: 0;
  height: 10vw;
  width: 100%;
`;

const SliderContainer = styled.div`
  // visibility: hidden;

  // ${({ imagesLoaded }) => imagesLoaded && `
  //     visibility: visible;
  // `}
`;

const LocoRow = (
  {
    setModalProps,
    numSlidersLoaded,
    mediaType,
    genreName,
    genreId,
    handleItemLoaded,
  },
) => {
  

  return (
    <RowContainer>
      <RowPadding />
      <RowHeader>
        <HeaderText>
          {genreName}
        </HeaderText>
      </RowHeader>
      <Row>
        {/* {!imagesLoaded &&
          <LoadingSkeleton />
        } */}
        <SliderContainer>
          <MediaSliderContainer
            setModalProps={setModalProps}
            numSlidersLoaded={numSlidersLoaded}
            mediaType={mediaType}
            genreName={genreName}
            genreId={genreId}
            handleItemLoaded={handleItemLoaded}
          />
        </SliderContainer>
      </Row>
      <RowPadding />
    </RowContainer>
  );
};

export default LocoRow;
