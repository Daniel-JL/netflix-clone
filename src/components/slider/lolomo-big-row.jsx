/* eslint-disable import/prefer-default-export */
import React from 'react';
import {
  Link,
} from 'react-router-dom';
import styled from 'styled-components';

const BigRowContainer = styled.div`
  width: 100%;
  padding-top: 41%;
  border: 1px solid black; 
`;

export function LolomoBigRow() {
  return (
    <BigRowContainer>

    </BigRowContainer>
  );
}
