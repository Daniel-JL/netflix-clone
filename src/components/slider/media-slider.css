.slides {
  margin: 60px;
  .slick-prev, .slick-next {
    top: 50%;
  }
  .slick-prev {
    z-index: 1;
    background-color: black;
  }
  .slick-next:before {
    z-index: 5;
  }
  .slick-track {
    display: flex;
  }
  .slick-track .slick-slide {
    display: flex;
    height: auto;
    align-items: center;
    justify-content: center;
  }
  .slick-slide {
    margin: 0 27px;
  }
  .slick-list {
    margin: 0 -27px;
  }

}

