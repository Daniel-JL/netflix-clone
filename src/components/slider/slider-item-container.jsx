import React, {
  useState, useRef, useEffect, createContext,
} from 'react';
import { SliderItem } from './slider-item';
import { getMediaData } from '../../helpers/getMediaData';
import { getAgeRating } from '../../helpers/getAgeRating';
import mediaIsMovie from '../../helpers/mediaIsMovie';

const SliderItemContainer = ({
  setModalProps,
  mediaType,
  mediaId,
  handleImageLoaded,
}) => {
  const [dataLoaded, setDataLoaded] = useState(false);
  const [imgLoadingErr, setImgLoadingErr] = useState(false);
  const [imgLoadedSuccess, setImgLoadedSuccess] = useState(false);
  const [itemHoverActive, setItemHoverActive] = useState(false);
  const [itemHoverTransition, setItemHoverTransition] = useState(true);
  const [itemDimensions, setItemDimensions] = useState({});
  const [mediaTitle, setMediaTitle] = useState();
  const [posterPath, setPosterPath] = useState();
  const [genres, setGenres] = useState();
  const [runtimeOrNumberOfSeasons, setRuntimeOrNumberOfSeasons] = useState();
  const [ageRating, setAgeRating] = useState();
  const [overview, setOverview] = useState();
  const [delayHandler, setDelayHandler] = useState();
  const delayDuration = 300;

  const fetchItemData = async () => {
    const data = await getMediaData(mediaType, mediaId);

    setMediaTitle((mediaTitle) => (data.title ? data.title : data.name));
    setPosterPath((posterPath) => `http://image.tmdb.org/t/p/w780${data.backdrop_path}`);
    setGenres((genres) => [
      ...Array(data.genres.length),
    ].map((undefined, index) => data.genres[index].name));
    setRuntimeOrNumberOfSeasons((runtimeOrNumberOfSeasons) => (mediaIsMovie(mediaType) ? `${data.runtime}m` : data.number_of_seasons));
    setOverview((overview) => data.overview);

    const ageRating = await getAgeRating(mediaId, mediaType);
    setAgeRating((ageRating) => ageRating);
  

    setDataLoaded(true);
  };

  const handleMouseOver = (itemContainerRef) => {
    if (imgLoadedSuccess) {
      setDelayHandler(setTimeout(() => {
        const rect = itemContainerRef.getBoundingClientRect();
        setItemDimensions((itemDimensions) => ({
          left: rect.x,
          top: rect.y + window.scrollY,
          bottom: rect.bottom,
          right: rect.right,
          width: rect.right - rect.left,
          height: rect.bottom - rect.top,
        }));
        setItemHoverTransition(true);
        setItemHoverActive(true);
      }, delayDuration));
    }
  };

  const handleMouseOut = () => {
    clearTimeout(delayHandler);
    setItemHoverTransition(false);
  };

  const handleModalDismount = () => {
    setItemHoverActive(false);
  };

  const handleImgLoadingErr = () => {
    setImgLoadingErr(true);
  };

  const handleImgLoadedSuccess = () => {
    setImgLoadedSuccess(true);
    handleImageLoaded();
  };

  const handleEpsAndInfoButtonClick = () => {
    setModalProps(
      mediaId,
      mediaType,
      posterPath,
      runtimeOrNumberOfSeasons,
      genres,
      ageRating,
      overview,

    );
  };

  useEffect(() => {
    if (!dataLoaded) {
      fetchItemData();
    }
  });

  return (
    <div>
      {dataLoaded
        && (
        <SliderItem
          mediaTitle={mediaTitle}
          mediaId={mediaId}
          mediaType={mediaType}
          ageRating={ageRating}
          genres={genres}
          imgLoadedSuccess={imgLoadedSuccess}
          imgLoadingErr={imgLoadingErr}
          dataLoaded={dataLoaded}
          posterPath={posterPath}
          runtimeOrNumberOfSeasons={runtimeOrNumberOfSeasons}
          itemHoverActive={itemHoverActive}
          itemHoverTransition={itemHoverTransition}
          handleMouseOver={handleMouseOver}
          handleMouseOut={handleMouseOut}
          handleModalDismount={handleModalDismount}
          setImgLoadingErr={handleImgLoadingErr}
          setImgLoadSuccess={handleImgLoadedSuccess}
          handleEpsAndInfoButtonClick={handleEpsAndInfoButtonClick}
          itemDimensions={itemDimensions}
        />
        )}
    </div>
  );
};

export default SliderItemContainer;
