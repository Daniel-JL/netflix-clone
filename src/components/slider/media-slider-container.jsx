/* eslint-disable import/prefer-default-export */
import React, { useState, useRef, useEffect } from 'react';
import styled from 'styled-components';
import { MediaSlider } from './media-slider';
import getTrendingMediaIdsAndTypes from '../../helpers/getTrendingMediaIdsAndTypes';
import getMediaListByGenre from '../../helpers/getMediaListByGenre';
import processIdsAndTypes from '../../helpers/processIdsAndTypes';

const Container = styled.div`
  position: relative;
  z-index: 2;
  width: 100%;
`;

export function MediaSliderContainer(
  {
    setModalProps,
    numSlidersLoaded,
    mediaType,
    genreName,
    genreId,
    changeRowZIndex,
  },
) {
  const [paginationPage, setPaginationPage] = useState(0);
  const [buttonHasBeenPressed, setButtonHasBeenPressed] = useState(false);
  const [numOfVisibleSliderItems, setNumOfVisibleSliderItems] = useState(6);
  const [dataLoaded, setDataLoaded] = useState(false);
  const maxIdsNeeded = 42;
  const mediaIdsAndTypes = useRef({});

  const fetchNow = async () => {
    let data;
    if (genreName === 'trending') {
      data = await getTrendingMediaIdsAndTypes(maxIdsNeeded, numSlidersLoaded, mediaType);
      mediaIdsAndTypes.current = processIdsAndTypes(data, maxIdsNeeded);
    } else {
      const data = await getMediaListByGenre(mediaType, numSlidersLoaded, genreId, maxIdsNeeded);
      mediaIdsAndTypes.current = processIdsAndTypes(data, maxIdsNeeded, mediaType);
    }
    if (mediaIdsAndTypes.current.ids.length >= maxIdsNeeded) {
      setDataLoaded(true);
    }
  };

  useEffect(() => {
    if (!dataLoaded) {
      fetchNow();
    }
  }, []);

  return (
    <Container>
      {dataLoaded
        && (
        <MediaSlider
          setModalProps={setModalProps}
          dataLoaded={dataLoaded}
          maxIdsNeeded={maxIdsNeeded}
          movieTvShowType={mediaIdsAndTypes.current.types}
          movieTvShowIds={mediaIdsAndTypes.current.ids}
          changeRowZIndex={changeRowZIndex}
        />
        )}

    </Container>
  );
}
