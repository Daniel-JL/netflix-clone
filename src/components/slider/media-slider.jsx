import React, { useState, useRef, useEffect } from 'react';
import styled from 'styled-components';
import Slider from 'react-slick';
import 'slick-carousel/slick/slick.css';
import 'slick-carousel/slick/slick-theme.css';
import './media-slider.css';

const SliderContainer = styled.div`
  position: relative;
  z-index: 2;
  max-width: 100%;
  width: 100%;
  color: white;
`;

const MediaSlider = (
  {
    children,
    dataLoaded,
  },
) => (
  <div>
    {dataLoaded
        && (
        <SliderContainer>
          <Slider
            arrows
            draggable={false}
            dots={false}
            infinite
            speed={500}
            slidesToScroll={6}
            slidesToShow={6}
            className="slides"
            centerMode={false}
            variableWidth={false}
            useTransform
            lazyLoad="ondemand"
            responsive={[
              {
                breakpoint: 1200,
                settings: {
                  slidesToShow: 5,
                  slidesToScroll: 5,
                  infinite: true,
                  dots: true,
                },
              },
              {
                breakpoint: 1000,
                settings: {
                  slidesToShow: 4,
                  slidesToScroll: 4,
                  initialSlide: 2,
                },
              },
              {
                breakpoint: 800,
                settings: {
                  slidesToShow: 3,
                  slidesToScroll: 3,
                  initialSlide: 3,
                },
              },
              {
                breakpoint: 600,
                settings: {
                  slidesToShow: 2,
                  slidesToScroll: 2,
                  initialSlide: 2,
                },
              },
              {
                breakpoint: 400,
                settings: {
                  slidesToShow: 1,
                  slidesToScroll: 1,
                },
              },
            ]}
          >
            {children}
          </Slider>
        </SliderContainer>
        )}
  </div>
);

export default MediaSlider;
