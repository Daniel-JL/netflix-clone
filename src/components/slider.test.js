import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Slider, fetchSliderItemIds } from './slider';
import 'isomorphic-fetch';

describe('Slider', () => {
  it('should render correctly', () => {
    const wrapper = renderer.create(<Slider />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it('should highlight slider item on mouseover after short delay', () => {

  });

  it('should scroll across to new set of slider items if arrow button is pressed', () => {

  });

  it('should add left arrow button once right arrow button has been pressed once', () => {

  });

  it('should update pagination indicator as arrow keys are pressed', () => {

  });

  it('should highlight arrow buttons on mouseover', () => {

  });

  it('should fetch 42 movie/tv show ids and store them in an array', () => {
    //  TODO
    //  1. Figure out how to mock fetch - normal fetch is too slow
    
    const data = fetchSliderItemIds();
    console.log(data);

    expect(data).toHaveLength(42);
  });
});
