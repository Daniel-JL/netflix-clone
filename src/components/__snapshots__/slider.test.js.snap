// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should render correctly 1`] = `
<div
  className="sc-fzpans gxFaOO"
>
  <div
    className="sc-fzplWN etOvms"
  />
  <div
    className="sc-AxjAm jrPXsu"
  />
  <div
    className="sc-AxjAm jrPXsu"
  />
  <div
    className="sc-AxjAm jrPXsu"
  />
  <div
    className="sc-AxjAm jrPXsu"
  />
  <div
    className="sc-AxjAm jrPXsu"
  />
  <div
    className="sc-AxjAm jrPXsu"
  />
  <div
    className="sc-fzplWN etOvms"
  >
    <button
      className="sc-fzoLsD bYBobW"
    />
  </div>
</div>
`;
